x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/StringUtils.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.eql.expression.function.scalar.string;
+
+import org.elasticsearch.common.Strings;
+
+import static org.elasticsearch.common.Strings.hasLength;
+
+final class StringUtils {
+
+    private StringUtils() {}
+
+    /**
+     * Returns a substring using the Python slice semantics, meaning
+     * start and end can be negative
+     */
+    static String substringSlice(String string, int start, int end) {
+        if (hasLength(string) == false) {
+            return string;
+        }
+
+        int length = string.length();
+        
+        // handle first negative values
+        while (start < 0) {
+            start += length;
+        }
+        while (end < 0) {
```

Comment:
Could this `while` be replaced with this simple math?

```
if (end < 0) {
   end = end - end/length * length + (end % length != 0 ? length : 0);
}
```

Reply:
No so simple :)

Reply:
:-) instead of subtracting length each time and then check if the value is still negative, the formula above subtracts the entire amount once by dividing the `end` value by `length` (to see how many lengths "fit" in "end") and taking the whole part only (not also the decimals), multiplying by length and adding a _possible_ remainder (defined as `end % length`).